### ISSUE TITLE
issues for multiple operations on persistent objects 

### ISSUE THREAD

**Issue by SyedElec** [OP-TEE/optee_os#798]
Hi,

I'm experiencing some problems when using multiple operations on persistent objects. 

I'm using optee_os 2.0.0 and qemu. 
I hadn't this problem on optee_os 1.0.0

I tried to write simple code to reproduce the issue on qemu because it happens randomly depending on the number of operation performed on one or multiple persistent objects and the platform used.

###### For example the code below panics at the 10h iteration

```c
TEE_ObjectHandle object = TEE_HANDLE_NULL;

int object_id = 0;

uint32_t flags =  TEE_DATA_FLAG_OVERWRITE | TEE_DATA_FLAG_ACCESS_WRITE_META;

int i = 0;

while (i < 20) {

	IMSG("\n\nLOOP : %d", i+1);

	TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE, &object_id, sizeof(int), flags, 
							   TEE_HANDLE_NULL, NULL, 0, &object);

	TEE_CloseAndDeletePersistentObject1(object);

	i++;
} 
```
###### Output :
```
LOOP : 10
ERROR:   TEE-CORE: ERROR:   USER-TA: Panic 0x0
ERROR:   TEE-CORE: TA panicked with code 0x0 usr_sp 0x1105f8 usr_lr 0x119529
DEBUG:   [0x0] TEE-CORE:user_ta_enter:521: tee_user_ta_enter: TA panicked with code 0x0
DEBUG:   [0x0] TEE-CORE:tee_ta_invoke_command:561:   => Error: ffff3024 of 3
```

###### The code below fails at the 9th iteration. 

```c
TEE_Result ret;

TEE_ObjectHandle object = TEE_HANDLE_NULL;

int object_id = 0;
TEE_ObjectInfo info;

char write_data[50] = {0};
uint32_t write_data_len = 50;
char read_data[50] = {0};
uint32_t read_bytes;

uint32_t flags =  TEE_DATA_FLAG_ACCESS_WRITE | TEE_DATA_FLAG_ACCESS_READ;
	
int i = 1;
TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE, &object_id, sizeof(int), flags, 
                           TEE_HANDLE_NULL, NULL, 0, &object);

while (i < 21) {

	IMSG("\n\nLOOP : %d", i);

	ret = TEE_SeekObjectData(object, 0, TEE_DATA_SEEK_SET);
	if (ret != TEE_SUCCESS) {
		EMSG("Seek failed : 0x%X", ret);
	}

	ret = TEE_WriteObjectData(object, write_data, write_data_len);
	if (ret != TEE_SUCCESS) {
		EMSG("Write failed : 0x%X", ret);
	}

	ret = TEE_SeekObjectData(object, 0, TEE_DATA_SEEK_SET);
	if (ret != TEE_SUCCESS) {
		EMSG("Seek failed : 0x%X", ret);
	}

	ret = TEE_ReadObjectData(object, read_data, write_data_len, &read_bytes);
	if (ret != TEE_SUCCESS) {
		EMSG("Read failed : 0x%X", ret);
	}

	i++;
}
```

The problem appears after a certain number of operations on the persistent object, for instance, if I add other operations like Read, Write or Seek, it will fail on one of these. 

###### Output I obtain:
```
ERROR:   TEE-CORE: Can not open meta file
ERROR:   TEE-CORE: Failed to open TEE file
ERROR:   TEE-CORE: Object corrupt
[...]
ERROR:   TEE-CORE: ERROR:   USER-TA: Seek failed : 0xF0100001
```

Thank you for your help.

**Comment by pascal-brand-st-dev** [comment]
I'm able to reproduce the problem. I'm looking at it.

**Comment by SyedElec** [comment]
@pascal-brand-st-dev Thank you.
I don't know if you already noticed but it seems that ```arg->ret``` returned by ```thread_rpc``` in [thread.c](https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/thread.c#L1229) corresponds to ```FFFF000C``` which means ```TEE_OUT_OF_MEMORY```

On QEMU, the return code ```FFFF000C``` appears on the 504th call (may vary on different platforms) of the function ```thread_rpc``` whatever the function regarding persistent objects used.

Apparently, all other functions related to persistent objects get through this and receive the ```TEE_OUT_OF_MEMORY``` error code.

**Comment by pascal-brand-st-dev** [comment]
I have created https://github.com/OP-TEE/optee_test/pull/96 to reproduce the issue inside optee_test.

From my investigations, there are multiple calls to `thread_rpc_alloc()` and `thread_rpc_free()` that look correctly balanced. These instruction looks correctly translated as `process_alloc()` and `process_free()` in the supplicant.
However,  in the driver, I can only observe `tee_shm_alloc()`, but no `tee_shm_free()`.

Unfortunately, I am not sure I will be able to look more before I'll be off. @jenswi-linaro @jforissier could you have a look?

**Comment by jforissier** [comment]
@pascal-brand-st-dev sure, I'll try it on HiKey tomorrow.

**Comment by jforissier** [comment]
The failure is reproducible on HiKey. I'm looking into it.

**Comment by jenswi-linaro** [comment]
Hopefully fixed in https://github.com/linaro-swg/linux/pull/10

**Comment by SyedElec** [comment]
@pascal-brand-st-dev @jforissier @jenswi-linaro Thank you very much.
Problem is fixed with [linaro-swg/linux#10](https://github.com/linaro-swg/linux/pull/10)
Closed issue.
